File: src/analysis/fit_distributions.py
"""
Distribution Fitting Analysis for CBB Prediction Model

Tests key metrics against common statistical distributions using MLE fitting,
KS goodness-of-fit tests, and AIC for model comparison.

Player-level stats are weighted by minutes played via resampling: each player's
observation is replicated proportionally to their minutes share, producing a
weighted sample that is then fit normally.

Usage:
    python src/analysis/fit_distributions.py
    python src/analysis/fit_distributions.py --table team_efficiency
    python src/analysis/fit_distributions.py --verbose
"""

import argparse
import sqlite3
import sys
import warnings
from pathlib import Path

import numpy as np
from scipy import stats

sys.path.insert(0, str(Path(__file__).parent.parent.parent))

# ── Distributions to test ────────────────────────────────────────────────────

DISTRIBUTIONS = {
    "normal": stats.norm,
    "lognormal": stats.lognorm,
    "gamma": stats.gamma,
    "beta": stats.beta,
    "skewnorm": stats.skewnorm,
    "t": stats.t,
    "weibull_min": stats.weibull_min,
    "exponential": stats.expon,
}

# Distributions that require positive data
POSITIVE_ONLY = {"lognormal", "gamma", "weibull_min", "exponential"}
# Distributions that require data in (0, 1)
UNIT_INTERVAL = {"beta"}


# ── Core fitting logic ───────────────────────────────────────────────────────

def fit_distribution(data: np.ndarray, dist_name: str, dist) -> dict | None:
    """
    Fit a single distribution to data via MLE and evaluate goodness-of-fit.

    Returns dict with params, ks_stat, p_value, aic, or None on failure.
    """
    try:
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            params = dist.fit(data)

            # KS test against fitted distribution
            ks_stat, p_value = stats.kstest(data, lambda x: dist.cdf(x, *params))

            # AIC = 2k - 2ln(L)
            log_likelihood = np.sum(dist.logpdf(data, *params))
            if not np.isfinite(log_likelihood):
                return None

            k = len(params)
            aic = 2 * k - 2 * log_likelihood

            return {
                "dist_name": dist_name,
                "params": params,
                "ks_stat": ks_stat,
                "p_value": p_value,
                "aic": aic,
                "log_likelihood": log_likelihood,
            }
    except Exception:
        return None


def fit_all_distributions(data: np.ndarray) -> list[dict]:
    """
    Fit all candidate distributions to data, returning results sorted by AIC.

    Automatically filters distribution candidates based on data range
    (e.g., skips lognormal for data with negative values).
    """
    data = data[np.isfinite(data)]
    if len(data) < 30:
        return []

    results = []
    data_min, data_max = data.min(), data.max()

    for dist_name, dist in DISTRIBUTIONS.items():
        # Skip distributions incompatible with data range
        if dist_name in POSITIVE_ONLY and data_min <= 0:
            continue
        if dist_name in UNIT_INTERVAL:
            if data_min <= 0 or data_max >= 1:
                continue

        result = fit_distribution(data, dist_name, dist)
        if result is not None:
            results.append(result)

    results.sort(key=lambda r: r["aic"])
    return results


def confidence_label(p_value: float) -> str:
    """Classify fit confidence based on KS test p-value."""
    if p_value >= 0.10:
        return "HIGH"
    elif p_value >= 0.01:
        return "MEDIUM"
    else:
        return "LOW"


# ── Data loading ─────────────────────────────────────────────────────────────

def load_team_efficiency(conn: sqlite3.Connection) -> dict[str, np.ndarray]:
    """Load team_efficiency metrics, keyed by column name."""
    metrics = [
        "adj_oe", "adj_de", "adj_tempo", "barthag", "overall_rank",
        "wins", "losses",
    ]
    result = {}
    for col in metrics:
        rows = conn.execute(
            f"SELECT {col} FROM team_efficiency WHERE {col} IS NOT NULL"
        ).fetchall()
        if rows:
            result[col] = np.array([r[0] for r in rows], dtype=float)
    return result


def load_player_stats_weighted(conn: sqlite3.Connection) -> dict[str, np.ndarray]:
    """
    Load player_stats metrics weighted by minutes played.

    Weighting is done via resampling: each player's observation is replicated
    proportionally to their minutes share (normalized so total sample size is
    preserved). Players with zero or null minutes are excluded.
    """
    metrics = [
        "ortg", "usage", "efg_pct", "ts_pct", "ftr",
        "ft_pct", "two_p_pct", "three_p_pct",
        "orb_pct", "drb_pct", "ast_pct", "to_pct", "blk_pct", "stl_pct",
        "bpm", "obpm", "dbpm",
        "pts_pg", "ast_pg", "treb_pg",
        "adj_oe", "adj_de",
    ]

    result = {}
    rng = np.random.default_rng(42)  # reproducible

    for col in metrics:
        rows = conn.execute(f"""
            SELECT {col}, minutes FROM player_stats
            WHERE {col} IS NOT NULL AND minutes IS NOT NULL AND minutes > 0
        """).fetchall()

        if not rows or len(rows) < 30:
            continue

        values = np.array([r[0] for r in rows], dtype=float)
        minutes = np.array([r[1] for r in rows], dtype=float)

        # Normalize minutes to weights summing to sample size
        weights = minutes / minutes.sum() * len(values)

        # Resample with replacement proportional to weights
        indices = rng.choice(len(values), size=len(values), replace=True, p=minutes / minutes.sum())
        result[col] = values[indices]

    return result


def load_team_bpi(conn: sqlite3.Connection) -> dict[str, np.ndarray]:
    """Load team_bpi metrics."""
    metrics = [
        "bpi", "bpi_offense", "bpi_defense", "bpi_7day_change",
        "sor_rank", "sos_past_rank",
        "wins", "losses", "top50_wins", "top50_losses",
    ]
    result = {}
    for col in metrics:
        rows = conn.execute(
            f"SELECT {col} FROM team_bpi WHERE {col} IS NOT NULL"
        ).fetchall()
        if rows and len(rows) >= 30:
            result[col] = np.array([r[0] for r in rows], dtype=float)
    return result


def load_game_metrics(conn: sqlite3.Connection) -> dict[str, np.ndarray]:
    """Load game-level metrics and derived fields."""
    result = {}

    # Score distributions
    for col in ["home_score", "away_score", "attendance"]:
        rows = conn.execute(
            f"SELECT {col} FROM games WHERE {col} IS NOT NULL"
        ).fetchall()
        if rows and len(rows) >= 30:
            result[col] = np.array([r[0] for r in rows], dtype=float)

    # Derived: score differential, total points
    rows = conn.execute("""
        SELECT home_score, away_score FROM games
        WHERE home_score IS NOT NULL AND away_score IS NOT NULL
    """).fetchall()
    if rows:
        home = np.array([r[0] for r in rows], dtype=float)
        away = np.array([r[1] for r in rows], dtype=float)
        result["score_differential"] = home - away
        result["total_points"] = home + away

    return result


# ── Reporting ────────────────────────────────────────────────────────────────

def print_results(table_name: str, metrics: dict[str, np.ndarray], verbose: bool = False):
    """Fit distributions for all metrics in a table and print results."""
    print(f"\n{'=' * 80}")
    print(f"  {table_name}")
    print(f"{'=' * 80}")
    print(f"{'Metric':<22} {'Best Fit':<14} {'KS Stat':>8} {'p-value':>10} {'Conf':>6}  {'Runner-up':<14} {'AIC Δ':>8}")
    print(f"{'-' * 22} {'-' * 14} {'-' * 8} {'-' * 10} {'-' * 6}  {'-' * 14} {'-' * 8}")

    low_confidence = []

    for col, data in sorted(metrics.items()):
        results = fit_all_distributions(data)
        if not results:
            print(f"{col:<22} {'(insufficient data)'}")
            continue

        best = results[0]
        conf = confidence_label(best["p_value"])
        runner_up = results[1]["dist_name"] if len(results) > 1 else "—"
        aic_delta = results[1]["aic"] - best["aic"] if len(results) > 1 else 0

        print(
            f"{col:<22} {best['dist_name']:<14} {best['ks_stat']:>8.4f} "
            f"{best['p_value']:>10.4f} {conf:>6}  {runner_up:<14} {aic_delta:>8.1f}"
        )

        if conf == "LOW":
            low_confidence.append((col, data, results))

        if verbose:
            print(f"  {'':>22} n={len(data)}, mean={data.mean():.2f}, std={data.std():.2f}, "
                  f"skew={stats.skew(data):.2f}, kurt={stats.kurtosis(data):.2f}")
            for r in results[:4]:
                marker = " <-- best" if r is best else ""
                print(f"  {'':>22} {r['dist_name']:<14} AIC={r['aic']:>12.1f}  "
                      f"KS={r['ks_stat']:.4f}  p={r['p_value']:.4f}{marker}")

    return low_confidence


def print_low_confidence_detail(low_confidence_items: list):
    """Print detailed analysis for metrics with low-confidence fits."""
    if not low_confidence_items:
        return

    print(f"\n{'=' * 80}")
    print("  LOW-CONFIDENCE FIT DETAILS")
    print(f"{'=' * 80}")

    for col, data, results in low_confidence_items:
        print(f"\n  {col} (n={len(data)})")
        print(f"    Summary: mean={data.mean():.3f}, median={np.median(data):.3f}, "
              f"std={data.std():.3f}")
        print(f"    Range: [{data.min():.3f}, {data.max():.3f}]")
        print(f"    Skewness: {stats.skew(data):.3f}, Kurtosis: {stats.kurtosis(data):.3f}")
        print(f"    Percentiles: 5th={np.percentile(data, 5):.3f}, "
              f"25th={np.percentile(data, 25):.3f}, "
              f"75th={np.percentile(data, 75):.3f}, "
              f"95th={np.percentile(data, 95):.3f}")
        print(f"    Top 5 fits:")
        for r in results[:5]:
            conf = confidence_label(r["p_value"])
            print(f"      {r['dist_name']:<14} KS={r['ks_stat']:.4f}  "
                  f"p={r['p_value']:.6f}  AIC={r['aic']:>12.1f}  [{conf}]")
            # Print distribution parameters in human-readable form
            param_names = _get_param_names(r["dist_name"])
            param_str = ", ".join(f"{n}={v:.4f}" for n, v in zip(param_names, r["params"]))
            print(f"        params: {param_str}")


def _get_param_names(dist_name: str) -> list[str]:
    """Return human-readable parameter names for a distribution."""
    names = {
        "normal": ["loc (μ)", "scale (σ)"],
        "lognormal": ["shape (σ)", "loc", "scale (eᵘ)"],
        "gamma": ["shape (a)", "loc", "scale"],
        "beta": ["a", "b", "loc", "scale"],
        "skewnorm": ["skew (a)", "loc", "scale"],
        "t": ["df", "loc", "scale"],
        "weibull_min": ["shape (c)", "loc", "scale"],
        "exponential": ["loc", "scale"],
    }
    default = [f"p{i}" for i in range(10)]
    return names.get(dist_name, default)


# ── Main ─────────────────────────────────────────────────────────────────────

def main():
    parser = argparse.ArgumentParser(description="Fit distributions to CBB metrics")
    parser.add_argument("--db", type=str, default=None, help="Database path")
    parser.add_argument("--table", type=str, default=None,
                        help="Only analyze a specific table")
    parser.add_argument("--verbose", action="store_true",
                        help="Show detailed fit results for all metrics")
    args = parser.parse_args()

    db_path = args.db or "data/cbb_prediction.db"
    conn = sqlite3.connect(db_path)

    all_low_confidence = []

    tables = {
        "team_efficiency": load_team_efficiency,
        "player_stats (minutes-weighted)": load_player_stats_weighted,
        "team_bpi": load_team_bpi,
        "games": load_game_metrics,
    }

    for table_name, loader in tables.items():
        if args.table and args.table not in table_name:
            continue
        metrics = loader(conn)
        low = print_results(table_name, metrics, verbose=args.verbose)
        all_low_confidence.extend(low)

    print_low_confidence_detail(all_low_confidence)

    conn.close()

    print(f"\n{'=' * 80}")
    print(f"  Done. {len(all_low_confidence)} metrics had LOW confidence fits.")
    if all_low_confidence:
        print("  Review the LOW-CONFIDENCE section above — these may need manual inspection")
        print("  or a non-standard distribution (mixture model, kernel density, etc.).")
    print(f"{'=' * 80}")


if __name__ == "__main__":
    main()

File: src/analysis/fit_distributions.R
#!/usr/bin/env Rscript
# =============================================================================
# Distribution Fitting Analysis for CBB Prediction Model
#
# Tests key metrics against common statistical distributions using MLE fitting,
# KS goodness-of-fit tests, and AIC for model comparison.
#
# Player-level stats are weighted by minutes played via resampling: each
# player's observation is replicated proportionally to their share of total
# minutes, producing a weighted sample that is then fit normally.
#
# Usage:
#   Rscript src/analysis/fit_distributions.R
#   Rscript src/analysis/fit_distributions.R --verbose
#   Rscript src/analysis/fit_distributions.R --table team_efficiency
# =============================================================================

suppressPackageStartupMessages({
  library(DBI)
  library(RSQLite)
  library(fitdistrplus)
  library(MASS)
})

set.seed(42)

# ── CLI args ─────────────────────────────────────────────────────────────────

args <- commandArgs(trailingOnly = TRUE)
VERBOSE    <- "--verbose" %in% args
TABLE_FILTER <- NULL
for (i in seq_along(args)) {
  if (args[i] == "--table" && i < length(args)) {
    TABLE_FILTER <- args[i + 1]
  }
}

DB_PATH <- "data/cbb_prediction.db"

# ── Distribution candidates ──────────────────────────────────────────────────

# Each entry: list(name, fitdistrplus_name, requires_positive, requires_unit)
DISTRIBUTIONS <- list(
  list(name = "normal",      fname = "norm",    pos = FALSE, unit = FALSE),
  list(name = "lognormal",   fname = "lnorm",   pos = TRUE,  unit = FALSE),
  list(name = "gamma",       fname = "gamma",   pos = TRUE,  unit = FALSE),
  list(name = "beta",        fname = "beta",    pos = FALSE, unit = TRUE),
  list(name = "exponential", fname = "exp",     pos = TRUE,  unit = FALSE),
  list(name = "weibull",     fname = "weibull", pos = TRUE,  unit = FALSE),
  list(name = "logistic",    fname = "logis",   pos = FALSE, unit = FALSE),
  list(name = "t",           fname = "t",       pos = FALSE, unit = FALSE)
)

# ── Custom t-distribution support for fitdistrplus ───────────────────────────
# fitdistrplus needs dt/pt/qt functions with a single "df" shape parameter
# plus loc/scale. We define a 3-parameter t (df, loc, scale).

dt_ls <- function(x, df, loc = 0, scale = 1, log = FALSE) {
  z <- (x - loc) / scale
  if (log) {
    dt(z, df, log = TRUE) - log(scale)
  } else {
    dt(z, df) / scale
  }
}

pt_ls <- function(q, df, loc = 0, scale = 1, lower.tail = TRUE, log.p = FALSE) {
  pt((q - loc) / scale, df, lower.tail = lower.tail, log.p = log.p)
}

qt_ls <- function(p, df, loc = 0, scale = 1, lower.tail = TRUE, log.p = FALSE) {
  qt(p, df, lower.tail = lower.tail, log.p = log.p) * scale + loc
}

# ── Core fitting logic ──────────────────────────────────────────────────────

fit_one_distribution <- function(data, dist_info) {
  # Returns NULL on failure, otherwise a list with fit results
  tryCatch({
    dname  <- dist_info$fname
    dlabel <- dist_info$name

    # Use custom t distribution
    if (dlabel == "t") {
      med  <- median(data)
      sc   <- mad(data, constant = 1.4826)
      if (sc == 0) sc <- sd(data)
      start <- list(df = 5, loc = med, scale = sc)
      lower <- c(1.01, -Inf, 0.001)

      fit <- fitdist(data, "t_ls",
                     start = start, lower = lower,
                     method = "mle", keepdata = FALSE)

      params <- as.list(fit$estimate)
      ks     <- suppressWarnings(
        ks.test(data, "pt_ls",
                df = params$df, loc = params$loc, scale = params$scale)
      )

      return(list(
        name       = dlabel,
        aic        = fit$aic,
        ks_stat    = unname(ks$statistic),
        p_value    = ks$p.value,
        params     = params,
        loglik     = fit$loglik,
        fit_obj    = fit
      ))
    }

    # Standard distributions via fitdistrplus
    fit <- fitdist(data, dname, method = "mle", keepdata = FALSE)

    # KS test against fitted CDF
    params <- as.list(fit$estimate)
    pfun   <- match.fun(paste0("p", dname))
    ks     <- suppressWarnings(
      do.call(ks.test, c(list(x = data), pfun, params))
    )

    list(
      name       = dlabel,
      aic        = fit$aic,
      ks_stat    = unname(ks$statistic),
      p_value    = ks$p.value,
      params     = params,
      loglik     = fit$loglik,
      fit_obj    = fit
    )
  }, error = function(e) NULL)
}


fit_all_distributions <- function(data, label = "") {
  # Remove non-finite values
  data <- data[is.finite(data)]
  if (length(data) < 30) {
    return(list(results = list(), data = data))
  }

  data_min <- min(data)
  data_max <- max(data)

  results <- list()

  for (dist_info in DISTRIBUTIONS) {
    # Skip incompatible distributions
    if (dist_info$pos && data_min <= 0) next
    if (dist_info$unit && (data_min <= 0 || data_max >= 1)) next

    res <- fit_one_distribution(data, dist_info)
    if (!is.null(res)) {
      results[[length(results) + 1]] <- res
    }
  }

  # Sort by AIC
  if (length(results) > 0) {
    aics <- sapply(results, function(r) r$aic)
    results <- results[order(aics)]
  }

  list(results = results, data = data)
}


confidence_label <- function(p_value) {
  if (is.na(p_value)) return("  N/A")
  if (p_value >= 0.10) return(" HIGH")
  if (p_value >= 0.01) return("  MED")
  return("  LOW")
}


# ── Data loading ─────────────────────────────────────────────────────────────

load_team_efficiency <- function(conn) {
  metrics <- c("adj_oe", "adj_de", "adj_tempo", "barthag",
               "overall_rank", "wins", "losses")
  result <- list()
  for (col in metrics) {
    sql <- sprintf("SELECT %s FROM team_efficiency WHERE %s IS NOT NULL", col, col)
    vals <- dbGetQuery(conn, sql)[[1]]
    if (length(vals) >= 30) result[[col]] <- as.numeric(vals)
  }
  result
}


load_player_stats_weighted <- function(conn) {
  metrics <- c("ortg", "usage", "efg_pct", "ts_pct", "ftr",
               "ft_pct", "two_p_pct", "three_p_pct",
               "orb_pct", "drb_pct", "ast_pct", "to_pct", "blk_pct", "stl_pct",
               "bpm", "obpm", "dbpm",
               "pts_pg", "ast_pg", "treb_pg",
               "adj_oe", "adj_de")

  result <- list()

  for (col in metrics) {
    sql <- sprintf(
      "SELECT %s, minutes FROM player_stats
       WHERE %s IS NOT NULL AND minutes IS NOT NULL AND minutes > 0",
      col, col
    )
    df <- dbGetQuery(conn, sql)
    if (nrow(df) < 30) next

    values  <- as.numeric(df[[1]])
    minutes <- as.numeric(df$minutes)

    # Weighted resampling: resample proportional to minutes share
    probs   <- minutes / sum(minutes)
    indices <- sample.int(length(values), size = length(values),
                          replace = TRUE, prob = probs)
    result[[col]] <- values[indices]
  }
  result
}


load_team_bpi <- function(conn) {
  metrics <- c("bpi", "bpi_offense", "bpi_defense", "bpi_7day_change",
               "sor_rank", "sos_past_rank",
               "wins", "losses", "top50_wins", "top50_losses")
  result <- list()
  for (col in metrics) {
    sql <- sprintf("SELECT %s FROM team_bpi WHERE %s IS NOT NULL", col, col)
    vals <- dbGetQuery(conn, sql)[[1]]
    if (length(vals) >= 30) result[[col]] <- as.numeric(vals)
  }
  result
}


load_game_metrics <- function(conn) {
  result <- list()

  for (col in c("home_score", "away_score", "attendance")) {
    sql <- sprintf("SELECT %s FROM games WHERE %s IS NOT NULL", col, col)
    vals <- dbGetQuery(conn, sql)[[1]]
    if (length(vals) >= 30) result[[col]] <- as.numeric(vals)
  }

  # Derived metrics
  df <- dbGetQuery(conn,
    "SELECT home_score, away_score FROM games
     WHERE home_score IS NOT NULL AND away_score IS NOT NULL")
  if (nrow(df) > 0) {
    result[["score_differential"]] <- as.numeric(df$home_score - df$away_score)
    result[["total_points"]]       <- as.numeric(df$home_score + df$away_score)
  }
  result
}


# ── Reporting ────────────────────────────────────────────────────────────────

print_header <- function(table_name) {
  cat("\n")
  cat(strrep("=", 90), "\n")
  cat("  ", table_name, "\n")
  cat(strrep("=", 90), "\n")
  cat(sprintf("%-22s %-14s %8s %10s %5s  %-14s %8s\n",
              "Metric", "Best Fit", "KS Stat", "p-value", "Conf",
              "Runner-up", "AIC Dlt"))
  cat(sprintf("%-22s %-14s %8s %10s %5s  %-14s %8s\n",
              strrep("-", 22), strrep("-", 14), strrep("-", 8),
              strrep("-", 10), strrep("-", 5), strrep("-", 14), strrep("-", 8)))
}


print_results <- function(table_name, metrics) {
  print_header(table_name)
  low_confidence <- list()

  for (col in sort(names(metrics))) {
    data <- metrics[[col]]
    out  <- fit_all_distributions(data, col)
    results <- out$results

    if (length(results) == 0) {
      cat(sprintf("%-22s (insufficient data or fit failure)\n", col))
      next
    }

    best      <- results[[1]]
    conf      <- confidence_label(best$p_value)
    runner_up <- if (length(results) > 1) results[[2]]$name else "---"
    aic_delta <- if (length(results) > 1) results[[2]]$aic - best$aic else 0

    cat(sprintf("%-22s %-14s %8.4f %10.4f %5s  %-14s %8.1f\n",
                col, best$name, best$ks_stat, best$p_value, conf,
                runner_up, aic_delta))

    if (VERBOSE) {
      cat(sprintf("  %22s n=%d, mean=%.2f, sd=%.2f, skew=%.2f, kurt=%.2f\n",
                  "",
                  length(data), mean(data), sd(data),
                  moments_skew(data), moments_kurt(data)))
      for (r in results[1:min(4, length(results))]) {
        marker <- if (identical(r, best)) " <-- best" else ""
        cat(sprintf("  %22s %-14s AIC=%12.1f  KS=%.4f  p=%.4f%s\n",
                    "", r$name, r$aic, r$ks_stat, r$p_value, marker))
      }
    }

    if (trimws(conf) == "LOW") {
      low_confidence[[length(low_confidence) + 1]] <- list(
        col = col, data = data, results = results
      )
    }
  }

  low_confidence
}


print_low_confidence <- function(low_items) {
  if (length(low_items) == 0) return()

  cat("\n")
  cat(strrep("=", 90), "\n")
  cat("  LOW-CONFIDENCE FIT DETAILS\n")
  cat(strrep("=", 90), "\n")

  for (item in low_items) {
    d <- item$data
    cat(sprintf("\n  %s (n=%d)\n", item$col, length(d)))
    cat(sprintf("    Summary: mean=%.3f, median=%.3f, sd=%.3f\n",
                mean(d), median(d), sd(d)))
    cat(sprintf("    Range: [%.3f, %.3f]\n", min(d), max(d)))
    cat(sprintf("    Skewness: %.3f, Kurtosis: %.3f\n",
                moments_skew(d), moments_kurt(d)))
    pcts <- quantile(d, probs = c(0.05, 0.25, 0.75, 0.95))
    cat(sprintf("    Percentiles: 5th=%.3f, 25th=%.3f, 75th=%.3f, 95th=%.3f\n",
                pcts[1], pcts[2], pcts[3], pcts[4]))
    cat("    Top 5 fits:\n")
    for (r in item$results[1:min(5, length(item$results))]) {
      conf <- confidence_label(r$p_value)
      cat(sprintf("      %-14s KS=%.4f  p=%.6f  AIC=%12.1f  [%s]\n",
                  r$name, r$ks_stat, r$p_value, r$aic, trimws(conf)))
      pstr <- paste(
        mapply(function(n, v) sprintf("%s=%.4f", n, v),
               names(r$params), unlist(r$params)),
        collapse = ", ")
      cat(sprintf("        params: %s\n", pstr))
    }
  }
}


# ── Helper: moments ──────────────────────────────────────────────────────────

moments_skew <- function(x) {
  n  <- length(x)
  m  <- mean(x)
  s  <- sd(x)
  if (s == 0) return(0)
  (n / ((n - 1) * (n - 2))) * sum(((x - m) / s)^3)
}

moments_kurt <- function(x) {
  # Excess kurtosis (normal = 0)
  n <- length(x)
  m <- mean(x)
  s <- sd(x)
  if (s == 0) return(0)
  m4 <- mean((x - m)^4)
  m4 / s^4 - 3
}


# ── Main ─────────────────────────────────────────────────────────────────────

main <- function() {
  conn <- dbConnect(RSQLite::SQLite(), DB_PATH)
  on.exit(dbDisconnect(conn))

  all_low <- list()

  tables <- list(
    "team_efficiency"             = load_team_efficiency,
    "player_stats (min-weighted)" = load_player_stats_weighted,
    "team_bpi"                    = load_team_bpi,
    "games"                       = load_game_metrics
  )

  for (tname in names(tables)) {
    if (!is.null(TABLE_FILTER) && !grepl(TABLE_FILTER, tname, fixed = TRUE)) {
      next
    }
    metrics <- tables[[tname]](conn)
    low     <- print_results(tname, metrics)
    all_low <- c(all_low, low)
  }

  print_low_confidence(all_low)

  cat("\n")
  cat(strrep("=", 90), "\n")
  cat(sprintf("  Done. %d metrics had LOW confidence fits.\n", length(all_low)))
  if (length(all_low) > 0) {
    cat("  Review the LOW-CONFIDENCE section above -- these may need manual inspection\n")
    cat("  or a non-standard distribution (mixture model, KDE, etc.).\n")
  }
  cat(strrep("=", 90), "\n")
}

main()
